### PR TITLE
Added last_updated field to learner endpoints.

### PR DESCRIPTION
**Pull request by dsjen**
This introduces a new index that stories the last updated date for the learn roster entries index.

@dan-f, please review.

### CODE REVIEW

File: analytics_data_api/v0/serializers.py [hunk 1]
```diff
@@ -361,19 +362,6 @@ class EdxPaginationSerializer(pagination.PaginationSerializer):
     num_pages = serializers.Field(source='paginator.num_pages')
 
 
-class ElasticsearchDSLSearchSerializer(EdxPaginationSerializer):
```
Comment (dan-f) [inline analytics_data_api/v0/serializers.py]: Why did you remove this?

File: analytics_data_api/v0/views/learners.py [hunk 2]
```diff
@@ -223,7 +233,13 @@ def get_queryset(self):
         # parameter values in `get_users_in_course`.
         params = {key: val for key, val in params.items() if val is not None}
         try:
-            return RosterEntry.get_users_in_course(self.course_id, **params)
+            users = RosterEntry.get_users_in_course(self.course_id, **params)
+            last_updated = {'date': None}
+            if len(RosterUpdate.get_last_updated()) == 1:
+                last_updated = RosterUpdate.get_last_updated()[0]
```
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: Could you refactor this so that the `RosterUpdate.get_last_updated` query is only run once?

File: analytics_data_api/v0/views/learners.py [hunk 3]
```diff
@@ -223,7 +233,13 @@ def get_queryset(self):
...
+            else:
+                logger.warn('RosterUpdate not found.')
```
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: Is it possible for there to be no data for the requested course in the `ELASTICSEARCH_LEARNERS_UPDATE_INDEX`, but data in the `ELASTICSEARCH_LEARNERS_INDEX`?

File: analytics_data_api/v0/serializers.py [hunk 4]
```diff
@@ -426,20 +414,11 @@ def _get_range(self, metric_range):
 
 
 class CourseLearnerMetadataSerializer(serializers.Serializer):
-    enrollment_modes = serializers.SerializerMethodField('get_enrollment_modes')
-    segments = serializers.SerializerMethodField('get_segments')
-    cohorts = serializers.SerializerMethodField('get_cohorts')
+    enrollment_modes = serializers.Field(source='es_data.enrollment_modes')
```
Comment (dan-f) [inline analytics_data_api/v0/serializers.py]: :100: !!!!

File: analytics_data_api/v0/tests/views/test_learners.py [hunk 5]
```diff
@@ -87,6 +102,7 @@ def tearDown(self):
         """Remove the index after every test."""
         super(LearnerAPITestMixin, self).tearDown()
         self._es.indices.delete(index=settings.ELASTICSEARCH_LEARNERS_INDEX)
+        self._es.indices.delete(index=settings.ELASTICSEARCH_LEARNERS_UPDATE_INDEX)
```
Comment (dan-f) [inline analytics_data_api/v0/tests/views/test_learners.py]: I know it's not directly related to this PR, but would you mind refactoring this cleanup to use [`self.addCleanup`](https://docs.python.org/2/library/unittest.html#unittest.TestCase.addCleanup)? I recently learned that it's preferred over `self.tearDown`, since every statement in tearDown may not run if one of them fails.

File: analytics_data_api/v0/views/learners.py [hunk 6]
```diff
@@ -223,7 +233,13 @@ def get_queryset(self):
...
+            return [{'learner': user, 'last_updated': last_updated} for user in users.execute()]
```
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: I think I'd prefer to see this logic handled at the model layer. What do you think?

**Comment by dan-f** [comment]
I'm confused about the implementation here of the concept of "last updated". My understanding was that it would apply to the particular data that is being requested, but currently all endpoints will share the same last updated value. Is it true that "last updated" will be the same date across the board @mulby?

File: analytics_data_api/v0/views/learners.py (hunk 6 above)
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: Also, this implementation breaks the pagination values code path; for the pagination serializer to add `count`, `next`, `previous`, and `num_pages` it calls into the `elasticsearch_dsl` `Search` object. Now that we have a python list instead, `count` no longer is the number of matching results, `next` and `previous` are `null`, and `num_pages` is 1.

...why didn't the unit tests break?

**Comment by mulby** [comment]
@dan-f all data in the index should be updated at roughly the same time, so we save ourselves the overhead of an additional field on every record by just recording the time in one location. The timing doesn't need to be exact so this should be "good enough".

This should only apply to the elasticsearch data. Hopefully that's the case here.

File: analytics_data_api/v0/serializers.py (hunk 4 above)
Comment (dsjen) [inline analytics_data_api/v0/serializers.py]: :hamster: I can't tell you how excited I was to find this too!

File: analytics_data_api/v0/tests/views/test_learners.py (hunk 5 above)
Comment (dsjen) [inline analytics_data_api/v0/tests/views/test_learners.py]: That's very interesting!  I'll update this.

File: analytics_data_api/v0/views/learners.py (hunk 2 above)
Comment (dsjen) [inline analytics_data_api/v0/views/learners.py]: I don't mind updating this.  FYI, the call to `execute` is cached according to https://elasticsearch-dsl.readthedocs.org/en/latest/search_dsl.html.

File: analytics_data_api/v0/views/learners.py (hunk 3 above)
Comment (dsjen) [inline analytics_data_api/v0/views/learners.py]: The intension is for for update index to complement the learn index, and my thinking here is that it doesn't make sense for the entire call to fail if the update index isn't available for some reason.  What do you think?

File: analytics_data_api/v0/views/learners.py (hunk 6 above)
Comment (dsjen) [inline analytics_data_api/v0/views/learners.py]: Oh!  I tested this out quickly and saw that we were getting results for those fields (empty it turns out) and thought everything was working as expected.

I agree about moving the `execute()` call to the model--actually had intended on doing that.  I'll bring back `ElasticsearchDSLSearchSerializer` and see if I can figure out what went awry with the tests; oddly, they passed.

File: analytics_data_api/v0/views/learners.py (hunk 2 above)
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: Good to know.

File: analytics_data_api/v0/views/learners.py (hunk 3 above)
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: I agree, but I wanted to bring it up just in case.

**Comment by dsjen** [comment]
@dan-f, I've pushed a commit to address your comments.  There didn't seem to be a way to make it all just work from within the serializer, so I ended up moving the `last_updated` field into its own serializer and updating the responses with the serialized data from the update index in the `list()` and `retrieve()` methods.

File: analytics_data_api/v0/tests/views/test_learners.py [hunk 7]
```diff
@@ -25,10 +25,16 @@ def setUp(self):
...
-        if self._es.indices.exists(index=settings.ELASTICSEARCH_LEARNERS_INDEX):
-            self._es.indices.delete(index=settings.ELASTICSEARCH_LEARNERS_INDEX)
-        self._es.indices.create(index=settings.ELASTICSEARCH_LEARNERS_INDEX)
+
+        for index in [settings.ELASTICSEARCH_LEARNERS_INDEX, settings.ELASTICSEARCH_LEARNERS_UPDATE_INDEX]:
+            # delete index if for some reason the index wasn't deleted
+            def delete_index(to_delete):
+                if self._es.indices.exists(index=to_delete):
+                    self._es.indices.delete(index=to_delete)
+            delete_index(index)
```
Comment (dan-f) [inline analytics_data_api/v0/tests/views/test_learners.py]: Why do we need to make this call in setup if the indexes are always deleted during cleanup?

File: analytics_data_api/v0/tests/views/test_learners.py [hunk 8]
```diff
@@ -141,6 +152,20 @@ def create_learners(self, learners):
             self._create_learner(**learner)
         self._es.indices.refresh(index=settings.ELASTICSEARCH_LEARNERS_INDEX)
 
+    def create_index_update(self, date=None):
```
Comment (dan-f) [inline analytics_data_api/v0/tests/views/test_learners.py]: Nit: rename to `create_update_index` since it's the update index?

File: analytics_data_api/v0/tests/views/test_learners.py [hunk 9]
```diff
@@ -172,8 +197,8 @@ def test_get_user(self, username, name, course_id, enrollment_mode, segments=Non
             "problem_attempts_per_completed": problem_attempts_per_completed,
             "attempt_ratio_order": attempt_ratio_order,
             "enrollment_date": enrollment_date,
-            "last_updated": last_updated,
         }])
+        self.create_index_update(last_updated)
```
Comment (dan-f) [inline analytics_data_api/v0/tests/views/test_learners.py]: Rather than call this in each test case, could you add it to `setUp`?

File: analytics_data_api/v0/views/learners.py [hunk 10]
```diff
@@ -186,11 +215,20 @@ def _validate_query_params(self):
...
                     raise ParameterValueError('Page size must be in the range [1, {}]'.format(self.max_paginate_by))
 
+    def list(self, request, *args, **kwargs):
+        """
+        Adds the last_updated field to the results.
+        """
+        response = super(LearnerListView, self).list(request, args, kwargs)
+        last_updated = self.get_last_updated()
+        for result in response.data['results']:
+            result.update(last_updated)
```
Comment (dan-f) [inline analytics_data_api/v0/views/learners.py]: This differs from the current API spec, where `last_updated` is a key in the top-level JSON object. However, I don't really mind this implementation, and it makes things more consistent. Would you mind updating the API spec?

File: analytics_data_api/v0/tests/views/test_learners.py (hunk 7 above)
Comment (dsjen) [inline analytics_data_api/v0/tests/views/test_learners.py]: I was debating whether to include it or not.  I decided to include it in case the index wasn't deleted before our new snazzy `addCleanup` functionality.  I'm thinking now that we can assume that the system is clean and that there isn't a need for calling `delete_index()` explicitly; I'll remove it.

File: analytics_data_api/v0/tests/views/test_learners.py (hunk 8 above)
Comment (dsjen) [inline analytics_data_api/v0/tests/views/test_learners.py]: Will do!

**Comment by dan-f** [comment]
Looking good! :+1: once my final comments are addressed.

File: analytics_data_api/v0/tests/views/test_learners.py (hunk 9 above)
Comment (dsjen) [inline analytics_data_api/v0/tests/views/test_learners.py]: Right on!

File: analytics_data_api/v0/views/learners.py (hunk 10 above)
Comment (dsjen) [inline analytics_data_api/v0/views/learners.py]: Thanks, updated!
